R/MsBackendCached.R

```diff
@@ -0,0 +1,616 @@
+    rep("<cache>", length(object))
+})
+
+#' @rdname MsBackendCached
+setMethod("length", "MsBackendCached", function(x) {
+    x@nspectra
+})
+
+#' @rdname MsBackendCached
+setMethod("spectraVariables", "MsBackendCached", function(object) {
+    unique(c(names(.SPECTRA_DATA_COLUMNS), colnames(object@localData),
+             object@spectraVariables))
+})
+
+#' @importFrom S4Vectors make_zero_col_DFrame
+#'
+#' @noRd
+.spectra_data <- function(x, columns = spectraVariables(x)) {
+    local_cols <- intersect(columns, colnames(x@localData))
+    ## @spectraVariables are supposed to be added by the extending backend
+    columns <- columns[!columns %in% x@spectraVariables]
+    core_cols <- intersect(columns, names(.SPECTRA_DATA_COLUMNS))
+    core_cols <- core_cols[!core_cols %in% c(local_cols, x@spectraVariables)]
+    res <- NULL
+    if (length(local_cols))
+        res <- as(x@localData[, local_cols, drop = FALSE], "DataFrame")
+    if (length(core_cols)) {
+        mzvals <- NULL
+        intvals <- NULL
+        lst <- NumericList(numeric(), compress = FALSE)
+        if (any(core_cols == "mz")) {
+            core_cols <- core_cols[core_cols != "mz"]
+            mzvals <- lst[rep(1, times = length(x))]
+        }
+        if (any(core_cols == "intensity")) {
+            core_cols <- core_cols[core_cols != "intensity"]
+            intvals <- lst[rep(1, times = length(x))]
+        }
+        if (length(core_cols))
+            tmp <- DataFrame(lapply(.SPECTRA_DATA_COLUMNS[core_cols],
+                                    function(z, n) rep(as(NA, z), n), length(x)))
+        else tmp <- make_zero_col_DFrame(x@nspectra)
+        tmp$mz <- mzvals
+        tmp$intensity <- intvals
+        if (length(res))
+            res <- cbind(res, tmp)
+        else res <- tmp
+        if (any(core_cols == "dataStorage"))
+            res$dataStorage <- dataStorage(x)
+    }
+    if (!all(columns %in% colnames(res)))
+        stop("Column(s) ", paste0(columns[!columns %in% names(res)],
+                                  collapse = ", "), " not available.",
+             call. = FALSE)
+    if (length(res))
+        as(res, "DataFrame")
+    else NULL
+}
+
+#' @rdname MsBackendCached
+setMethod(
+    "spectraData", "MsBackendCached",
+    function(object, columns = spectraVariables(object)) {
+        .spectra_data(object, columns = columns)
+    })
+
+#' @rdname MsBackendCached
+setReplaceMethod("spectraData", "MsBackendCached",function(object, value) {
+    stop(class(object)[1], " does not support replacing the full spectra data.")
+})
+
+#' @rdname MsBackendCached
+setMethod("[", "MsBackendCached", function(x, i, j, ..., drop = FALSE) {
+    if (missing(i))
+        return(x)
+    slot(x, "localData", check = FALSE) <- x@localData[i, , drop = FALSE]
+    x@nspectra <- nrow(x@localData)
+    x
+})
+
+#' @rdname MsBackendCached
+setMethod("$", "MsBackendCached", function(x, name) {
+    if (!any(spectraVariables(x) == name))
```

Comment:
```suggestion
    if (all(spectraVariables(x) != name))
```

Reply:
Isn't `any` faster than `all` because it breaks when the first `TRUE` is returned?

Reply:
Yes, I guess you are right. I find `all( != )` easier to read and thought it would not be a time critical step. But you can keep `any`.